src/nvim/ex_docmd.c

```diff
@@ -7802,11 +7802,59 @@ static void ex_redraw(exarg_T *eap)
   if (eap->forceit) {
     redraw_all_later(NOT_VALID);
   }
-  update_screen(eap->forceit ? NOT_VALID
-                : VIsual_active ? INVERTED : 0);
+
+  // When NOT_VALID windows are redrawn, they lose incsearch highlighting since
+  // the last search pattern is reset.
+  // With display+=msg_sep, windows covered by the messages are NOT_VALID.
+  int need_redraw_highlight = (State & CMDLINE)
+      && ((p_is
+           && msg_scrolled > 0
+           && (dy_flags & DY_MSGSEP))
```

Comment:
Are we sure this is only for msgsep? without msgsep, instead the top rows of the screen will be invalidated and redrawn, can't these also need incsearch redraw?

Reply:
For the case with `display-=msgsep`, the windows which are above the messages have `w_redr_status` set to `REDRAW_TOP`, which appears to not clear the `incsearch` highlighting. 

You are correct regarding the other windows, they will lose the `incsearch` highlighting as well. I think this means `p_is && msg_scrolled > 0` should be the condition.